discovery_engine_core/core/src/storage/sqlite.rs

```diff
@@ -60,74 +59,99 @@ impl SqliteStorage {
         Ok(Self { pool })
     }
 
-    async fn store_new_documents<'a>(
-        mut tx: Transaction<'a, Sqlite>,
-        documents: impl Iterator<Item = &NewDocument> + Clone + Send,
-    ) -> Result<Transaction<'a, Sqlite>, Error> {
-        // insert id into Document table (fk of HistoricDocument)
-        let mut query_builder = QueryBuilder::new("INSERT INTO Document (id) ");
-        query_builder.push_values(documents.clone(), |mut stm, doc| {
-            stm.push_bind(doc.id.as_uuid());
-        });
+    async fn begin_tx(&self) -> Result<Transaction<'_, Sqlite>, Error> {
+        self.pool
+            .begin()
+            .await
+            .map_err(|err| Error::Database(err.into()))
+    }
+
+    async fn commit_tx(tx: Transaction<'_, Sqlite>) -> Result<(), Error> {
+        tx.commit().await.map_err(|err| Error::Database(err.into()))
+    }
+
+    async fn store_new_documents<'a, I>(
+        tx: &mut Transaction<'_, Sqlite>,
+        documents: I,
+    ) -> Result<(), Error>
+    where
+        I: IntoIterator<Item = &'a NewDocument> + Send,
+        <I as IntoIterator>::IntoIter: Clone,
+    {
+        let documents = documents.into_iter();
+        if documents.clone().next().is_none() {
+            return Ok(());
```

Comment:
❔ shouldn't we return some type of error in this case, indicating that the operation did not happened?

Reply:
this is an optimization, as (ignoring unspecified side effects) the difference between running a SQL query which stores 0 elements and not doing anything is non-observable.

On the other hand if we would error we also would need to potentially special case all code which maybe could end up with an empty list of documents it want's to store. Which isn't good, as it's more unnecessary code and more prone to lead to errors on the long run.

So in general if you want to delete, change, insert exactly 0 things doing nothing and just returning ok is preferable.